File: 90. 子集 II/90.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<int> tmp;
        sort(nums.begin(),nums.end());
        helper(nums,tmp);
        return res;
    }

    void helper(vector<int> nums,vector<int> tmp)
    {
        res.push_back(tmp);
        for(unsigned i=0;i<nums.size();i++)
        {
            if(i>0 && nums[i]==nums[i-1])
                continue;
            vector<int> new_nums(nums.begin()+i+1,nums.end());
            tmp.push_back(nums[i]);
            helper(new_nums,tmp);
            tmp.pop_back();
        }
    }
};

//翻译了一遍python代码，注意因为tmp放在外面了，所以最后要pop

File: 90. 子集 II/90.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        def backtrack(nums,tmp):
            res.append(tmp)
            for i in range(len(nums)):
                if(i-1>=0 and nums[i-1]==nums[i]):
                    continue
                backtrack(nums[i+1:],tmp+[nums[i]])
        backtrack(nums,[])
        return res

# 一个简单的回溯，之前写的很多，现在都写不出来了